### PR TITLE
Commit author + Serverless

### PR DESCRIPTION
**Pull request by naseemkullah**
Hi @Philmod ,

The objective of this commit is two fold:
- Have the slack message mention the github commit author if `GITHUB_TOKEN` is defined
- Add the option of using the serverless framework to manage the cloud function

Along with these 2 major features, I've linted, updated packages and renamed the package to `google-cloud-build-slack`

Many thanks to @iliasbhal and @juanborre for their help on this.

There is a vital piece missing atm, determining how to retrieve the github OWNER and REPO values, at this time it is assumed that Cloud Source Repository mirrors of github repos are in the form of `github_<OWNER>_<REPO>` which is not necessarily always the case.

TODO:
- [x] [~fool proof way to retrieve github OWNER and REPO~](https://cloud.google.com/cloud-build/docs/api/reference/rest/v1/projects.triggers#GitHubEventsConfig) mention of naming requirement for cloud source repo in order for github commit author info to be retrieved in README 
- [x] add serverless framework and github commit author instructions in README
- [x] Add unit tests

Finally, it is worth mentioning that the introduction of async/await in this commit has forced the use of `nodejs8` as the runtime

### CODE REVIEW

**Comment by Philmod** [comment]
Hey! Thanks for the PR. 
1. Could you send me a link to what you call "serveless"?
2. Can you write some unit tests?

**Comment by naseemkullah** [comment]
My pleasure!
1. https://serverless.com/framework/docs/providers/google/
2. Yes! I'll work on that.

File: index.js [hunk 1]
```diff
@@ -1,30 +1,36 @@
...
 
   // Send message to slack.
-  const message = module.exports.createSlackMessage(build);
-  module.exports.webhook.send(message, (err, res) => {
-    if (err) console.log('Error:', err);
-    callback(err);
+  const message = await module.exports.createSlackMessage(build);
+  module.exports.webhook.send(message, async (err, res) => {
+    if (err) {
+      console.log('Error:', err);
```
Comment (Philmod) [inline index.js]: The error should be surfaced to the caller

File: index.js [hunk 2]
```diff
@@ -90,11 +98,18 @@ module.exports.createSlackMessage = (build) => {
...
+        title: 'Commit Author',
+        value: commitAuthor,
+      });
+    }
   }
 
   // Add images to the message.
   const images = build.images || [];
-  for (let i = 0, len = images.length; i < len; i++) {
+  for (let i = 0, len = images.length; i < len; i += 1) {
```
Comment (Philmod) [inline index.js]: unnecessary change, please change back

File: index.js [hunk 3]
```diff
@@ -65,17 +75,15 @@ module.exports.createSlackMessage = (build) => {
...
-        footer_icon: 'https://ssl.gstatic.com/pantheon/images/containerregistry/container_registry_color.png',
-        ts: timestamp,
-      },
-    ],
+    attachments: [{
+      color: STATUS_COLOR[build.status] || DEFAULT_COLOR,
+      title: 'Build logs',
+      title_link: build.logUrl,
+      fields,
+      footer: 'Google Cloud Container Builder',
```
Comment (Philmod) [inline index.js]: 'Google Cloud Build'

File: index.js [hunk 4]
```diff
@@ -65,17 +75,15 @@ module.exports.createSlackMessage = (build) => {
...
-        color: STATUS_COLOR[build.status] || DEFAULT_COLOR,
-        title: 'Build logs',
-        title_link: build.logUrl,
-        fields,
-        footer: 'Google Cloud Container Builder',
-        footer_icon: 'https://ssl.gstatic.com/pantheon/images/containerregistry/container_registry_color.png',
-        ts: timestamp,
-      },
-    ],
+    attachments: [{
```
Comment (Philmod) [inline index.js]: unnecessary change, please change back
(try not to change code for the style)

File: index.js [hunk 5]
```diff
@@ -37,13 +43,17 @@ const STATUS_COLOR = {
...
 // createSlackMessage create a message from a build object.
-module.exports.createSlackMessage = (build) => {
+module.exports.createSlackMessage = async (build) => {
   const buildFinishTime = new Date(build.finishTime);
   const buildStartTime = new Date(build.startTime);
 
   const isWorking = build.status === 'WORKING';
   const timestamp = Math.round(((isWorking) ? buildStartTime : buildFinishTime).getTime() / 1000);
 
+  const commitAuthor = (token)
```
Comment (Philmod) [inline index.js]: Move this next to the code that uses it

File: lib/getCommitAuthor.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    const githubCommit = await octokit.git.getCommit({
+      commit_sha: commitSha,
+      owner: githubOwner,
+      repo: githubRepo,
+    });
+    // return github commit author e-mail
+    const commitAuthor = githubCommit.data.author.email;
+    return commitAuthor;
+  } catch (error) {
+    console.log(error);
```
Comment (Philmod) [inline lib/getCommitAuthor.js]: The error should be reported to the caller

**Review by Philmod — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (naseemkullah) [reply index.js]: refactored code, please review

File: lib/getCommitAuthor.js (hunk 6 above)
Comment (naseemkullah) [reply lib/getCommitAuthor.js]: refactored code, please review

**Comment by naseemkullah** [comment]
Hi @Philmod, I've added some unit tests, however I need to be able to unset `GITHUB_TOKEN` in a test to cover 2 more lines of code. I am researching on how to do that atm, if you have any suggestion please let me know.

**Comment by naseemkullah** [comment]
Hi @Philmod, unit tests have been added!

**Comment by naseemkullah** [comment]
Thanks @n8han for help on last commit (extracting dependency and passing down octokit as option param)

**Comment by naseemkullah** [comment]
bump

**Comment by naseemkullah** [comment]
bump

File: package.json [hunk 7]
```diff
@@ -11,17 +11,20 @@
...
-    "url": "git+https://github.com/Philmod/google-container-slack.git"
+    "url": "git+https://github.com/Philmod/google-cloud-build-slack.git"
   },
   "dependencies": {
-    "@slack/client": "4.8.0",
-    "humanize-duration": "3.17.0"
+    "@octokit/rest": "^16.20.0",
+    "@slack/client": "4.12.0",
+    "humanize-duration": "3.18.0",
+    "serverless-google-cloudfunctions": "^2.3.2"
```
Comment (Philmod) [inline package.json]: This should go to `devDependencies`

File: index.js [hunk 8]
```diff
@@ -36,8 +38,26 @@ const STATUS_COLOR = {
...
+    repo: githubRepo,
+  });
+
+  // return github commit
+  return githubCommit;
+};
+
 // createSlackMessage create a message from a build object.
-module.exports.createSlackMessage = (build) => {
+module.exports.createSlackMessage = async (build, octokit) => {
```
Comment (Philmod) [inline index.js]: It's a bit weird to pass `octokit` as an argument. 

2 options I can see: 
- Instead of a simple method, have a class on which octobit is attached
- Pass the `githubCommit` as argument instead, and have the parent function fetch it

WDYT?

File: index.js [hunk 9]
```diff
@@ -1,30 +1,32 @@
...
+const Octokit = require('@octokit/rest');
 const config = require('./config.json');
 
 module.exports.webhook = new IncomingWebhook(config.SLACK_WEBHOOK_URL);
 module.exports.status = config.GC_SLACK_STATUS;
 
 // subscribe is the main function called by GCF.
-module.exports.subscribe = (event, callback) => {
-  const build = module.exports.eventToBuild(event.data.data);
+module.exports.subscribe = async (event) => {
```
Comment (Philmod) [inline index.js:32]: With async, should the code wrapped in a try,catch so that we can return errors? 
Can you write a test in which something errors out?

**Review by Philmod — COMMENTED** [review]
I'm very sorry for responding so slowly, I've been very busy at work lately

**Comment by naseemkullah** [comment]
> I'm very sorry for responding so slowly, I've been very busy at work lately

No worries at all!!

File: index.js (hunk 8 above)
Comment (naseemkullah) [reply index.js]: True, it is weird indeed.

I've went with passing the `githubCommit` as argument instead, and having the parent function fetch it.
However the parent function still takes `octokit` as an argument... is that ok?

File: index.js (hunk 9 above)
Comment (naseemkullah) [reply index.js:32]: Try catch added and test written, please review!

**Comment by Philmod** [comment]
Thanks!

**Comment by naseemkullah** [comment]
Hi @Philmod , thanks for merging, I noticed this failed: https://github.com/Philmod/google-cloud-build-slack/runs/98545329
I cannot access the cloudbuild logs, if they cannot be made visible could you paste the error here please?

**Comment by naseemkullah** [comment]
https://github.com/Philmod/google-cloud-build-slack/pull/21 Created. Should fix the error
